src/lua/uri.c

```diff
@@ -0,0 +1,313 @@
+	if (param == NULL)
+		param = uri_add_query_param(uri, name);
+	int type = lua_type(L, -1);
+	switch (type) {
+	case LUA_TSTRING:
+	case LUA_TNUMBER:
+		uri_query_param_add_value(param, lua_tostring(L, -1));
+		break;
+	case LUA_TTABLE:
+		rc = uri_add_param_from_lua_table(param, L);
+		break;
+	default:
+		rc = -1;
+	}
+	return rc;
+}
+
+/**
+ * Add URI query parameter with @a name to all URIs in @a uri_set from the
+ * value at the top of the lua stack.
+ */
+static int
+uri_set_add_param_from_lua(struct uri_set *uri_set, struct lua_State *L,
+			   const char *name, bool if_not_exist)
+{
+	int rc = 0;
+	for (int i = 0; i < uri_set->uri_count && rc == 0; i++) {
+		struct uri *uri = &uri_set->uris[i];
+		if (uri_query_param_count(uri, name) != 0 && if_not_exist)
+			continue;
+		rc = uri_add_param_from_lua(uri, L, name);
+	}
+	return rc;
+}
+
+/**
+ * Add or overwrite (depends on @a param_is_overwritten) URI query parameters
+ * for all URIs in @a uri_set. Place the table appropriate to the @a table_name
+ * key on the top of the lua stack and use it to get the list of URI query
+ * parameters and there new values.
+ */
+static int
+uri_set_add_params_from_lua(struct uri_set *uri_set, struct lua_State *L,
+			    const char *table_name, bool param_is_overwritten)
+{
+	lua_pushstring(L, table_name);
+	lua_rawget(L, -2);
+	if (lua_type(L, -1) == LUA_TNIL) {
+		lua_pop(L, 1);
+		return 0;
+	} else if (lua_type(L, -1) != LUA_TTABLE) {
+		lua_pop(L, 1);
+		return -1;
+	}
+	int rc = 0;
+	lua_pushnil(L);
+	while (lua_next(L, lua_gettop(L) - 1) != 0 && rc == 0) {
+		if (lua_type(L, -2) == LUA_TSTRING) {
+			const char *param_name = lua_tostring(L, -2);
+			bool if_not_exist = !param_is_overwritten;
+			if (param_is_overwritten)
+				uri_set_clear_query_param(uri_set, param_name);
+			rc = uri_set_add_param_from_lua(uri_set, L, param_name,
+							if_not_exist);
+		} else {
+			rc = -1;
+		}
+		lua_pop(L, 1);
+	}
+	lua_pop(L, 1);
+	return rc;
+}
+
+/**
+ * Check if there is a field with the name @name in the table,
+ * which located at the top of the lua stack.
+ */
+static bool
+is_field_present(struct lua_State *L, const char *name)
+{
+	lua_pushstring(L, name);
+	lua_rawget(L, -2);
+	bool field_is_present = (lua_type(L, -1) != LUA_TNIL);
+	lua_pop(L, 1);
+	return field_is_present;
+}
+
+/**
+ * Check if there are nested tables in the table, which located
+ * at the top of the lua stack.
+ */
+static bool
+is_nested_uri_tables_present(struct lua_State *L)
+{
+	bool nested_tables_is_present = false;
+	int uri_count = lua_objlen(L, -1);
+	for (int i = 0; i < uri_count && !nested_tables_is_present; i++) {
+		lua_rawgeti(L, -1, i + 1);
+		nested_tables_is_present = (lua_type(L, -1) == LUA_TTABLE);
+		lua_pop(L, 1);
+	}
+	return nested_tables_is_present;
+}
+
+/**
+ * Perform some checks for the URI table. The types of checks depend on
+ * @a table_is_nested.
+ */
+static bool
+uri_table_is_valid(struct lua_State *L, bool table_is_nested)
```

Comment:
It's difficult to follow the checks when they are all mixed together in the same function. There should be a function that parses a single uri and does all related checks. Checks related to an array of uris should live in a function that creates a `uri_set` from a Lua object.

Reply:
fixed